### PR TITLE
net, evpn: fix EVPN test infrastructure for OVN-K VTEP BGP advertisement

### PR DESCRIPTION
**Pull request by servolkov**
##### What this PR does / why we need it:

Fix the external FRR pod nexthop advertisement to be compatible with
the VTEP host IP BGP advertisement introduced in recent OVN-K versions,
and strengthen the overall EVPN test setup.

**Main fix — frr-external nexthop advertisement:**
Recent OVN-K versions introduced per-node VTEP host IP advertisement
via BGP through the ovnk-generated FRRConfiguration. With eBGP
next-hop-self, the external FRR pod was re-advertising these /32 host
routes to peer nodes with itself as the next-hop, installing host routes
in the main routing table that redirected all inter-node traffic
(including etcd raft) through the FRR pod, which dropped it.
Fixed by using `attribute-unchanged next-hop` so the original node IP
is preserved, making the host routes harmless.

FRR reference:
https://docs.frrouting.org/en/latest/bgp.html#clicmd-neighbor-PEER-attribute-unchanged-as-path-next-hop-med

**EVPN setup improvements:**
- **use node-primary-ifaddr for node IP extraction**
Replace parsing of k8s.ovn.org/host-cidrs with k8s.ovn.org/node-primary-ifaddr.
OVN-K sets this annotation to the node's primary interface address per address
family, giving exactly one unambiguous entry per family.

- **ovn_local_gateway_mode yielded without waiting** 
Patching `routingViaHost: True` triggers an OVN-K DaemonSet rollout. 
Following the same pattern used in similar network operator patches 
across the repo (e.g. `cluster_network_resource_ra_enabled`, 
`multi_network_policy_enabled`), add `wait_for_consistent_resource_conditions` 
on the network ClusterOperator before yielding and after teardown. 
This is a safety measure to ensure the cluster is fully stable before and after 
applying the `routingViaHost: True` patch, consistent with how other fixtures 
handle network operator changes.

- **remove redundant as-path exclusion from evpn route-map**
OVN-K now generates `allowas-in origin` on nodes for both IPv4 unicast
and L2VPN EVPN address families, so cluster nodes accept routes where
their own ASN is the origin AS without needing the external FRR pod to
strip it first.
References:
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6307
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6481

##### jira-ticket:
NONE

## Bug Fixes

- Improved BGP external-router configuration to preserve IPv4 next-hop information and correctly handle EVPN routes.
- Improved EVPN gateway-mode updates by confirming network components reach consistent states before and after changes.

## Tests

- Improved EVPN and VTEP validation by using each worker’s primary IPv4 network information for more reliable network checks.

### CODE REVIEW

File: tests/network/bgp/evpn/libevpn.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
 import contextlib
 import ipaddress
+import json
 import logging
 import shlex
 import uuid
 from collections.abc import Generator
 from dataclasses import dataclass
 
+from ocp_resources.node import Node
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: fixed

File: tests/network/bgp/evpn/conftest.py [hunk 2]
```diff
@@ -88,9 +89,7 @@ def vtep(
     admin_client: DynamicClient,
     workers: list[Node],
 ) -> Generator[VTEP]:
-    host_cidrs = json.loads(workers[0].instance.metadata.annotations["k8s.ovn.org/host-cidrs"])
-    host_cidr = next(cidr for cidr in host_cidrs if "." in cidr)
-    vtep_cidr = str(ipaddress.ip_network(host_cidr, strict=False))
+    vtep_cidr = str(ipaddress.ip_network(node_primary_ipv4_cidr(workers[0]), strict=False))
```
Comment (qinqon) [inline tests/network/bgp/evpn/conftest.py]: why do we convert to ip address then back to str ?

File: tests/network/bgp/evpn/libevpn.py [hunk 3]
```diff
@@ -468,3 +473,14 @@ def assert_evpn_workloads_connectivity(
         for l3_client, l3_server in l3_connections:
             with subtests.test(f"routed-L3 IPv{ipaddress.ip_address(l3_client.server_ip).version}"):
                 assert is_tcp_connection(server=l3_server, client=l3_client)
+
+
+def node_primary_ipv4_cidr(node: Node) -> str:
```
Comment (qinqon) [inline tests/network/bgp/evpn/libevpn.py]: What about ipv6 or dual stack ? this is going to return ipv6 address with cidr < 32

Comment (qinqon) [reply tests/network/bgp/evpn/libevpn.py]: Also should we do something special for ipv6 ?

**Comment by servolkov** [comment]
@coderabbit review

File: tests/network/libs/bgp.py [hunk 4]
```diff
@@ -214,20 +214,19 @@ def generate_frr_conf(
...
     lines.extend([
         " address-family ipv4 unicast",
         f"  network {external_subnet_ipv4}",
     ])
     for ip in nodes_ipv4_list:
         lines.extend([
             f"  neighbor {ip} activate",
-            f"  neighbor {ip} next-hop-self",
-            f"  neighbor {ip} route-reflector-client",
+            f"  neighbor {ip} attribute-unchanged next-hop",
```
Comment (qinqon) [inline tests/network/libs/bgp.py:224]: as an alternative we can directly just remove those two lines, but this is what ovnk maintainers mentioned.

File: tests/network/bgp/evpn/libevpn.py [hunk 5]
```diff
@@ -1,12 +1,17 @@
 import contextlib
 import ipaddress
+import json
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py:3]: we dont do this way with std libs.

File: tests/network/libs/bgp.py
Comment (qinqon) [inline tests/network/libs/bgp.py]: set as-path exclude 64512 is no longer needed: ovn-kubernetes now generates neighbor <peer> allowas-in origin on the nodes for both unicast and l2vpn evpn, so they accept paths where 64512 is the origin AS. Removing it lets the test exercise the product's own eBGP handling instead of masking it.

- ovn-kubernetes/ovn-kubernetes#6307 — "EVPN: add allowas-in origin for BGP neighbors" (l2vpn evpn AF, merged 2026-04-28)
- ovn-kubernetes/ovn-kubernetes#6481 — "RA controller: add unicast allowas-in origin to raw FRR config" (ipv4/ipv6 unicast AF, merged 2026-06-10) — this is the one that made the exclude fully redundant

File: tests/network/bgp/evpn/libevpn.py [hunk 6]
```diff
@@ -468,3 +473,14 @@ def assert_evpn_workloads_connectivity(
...
                 assert is_tcp_connection(server=l3_server, client=l3_client)
+
+
+def node_primary_ipv4_cidr(node: Node) -> str:
+    """Return the primary IPv4 CIDR of a node, skipping /32 host routes.
+
+    Parses k8s.ovn.org/host-cidrs and returns the first non-/32 IPv4
+    CIDR entry. On clusters where /32 host-route entries appear before
+    the interface CIDR, this ensures the correct address is returned.
+    """
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: local helper, not necessary, the helper semantics is pretty clean.

**Review by qinqon — CHANGES_REQUESTED** [review]

File: tests/network/libs/bgp.py (hunk 4 above)
Comment (servolkov) [reply tests/network/libs/bgp.py:224]: yep, I decided to add for clarity.

File: tests/network/bgp/evpn/libevpn.py (hunk 3 above)
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: our nodes are dual-stack. But we filter IPs via `if "." in cidr`. So, how it can return IPv6?

>Also should we do something special for ipv6 ?

if you mean if have to setup ipv6 tunnel - it was not implemented in the product as far as I remember, we use only ipv4 VTEPs. Anyway, this is not related to the current PR.

File: tests/network/libs/bgp.py
Comment (servolkov) [reply tests/network/libs/bgp.py]: as we discussed offline you did not want to have this fix in this PR. I don't mind to have additional 1 line change, let me know.

File: tests/network/bgp/evpn/conftest.py (hunk 2 above)
Comment (servolkov) [reply tests/network/bgp/evpn/conftest.py]: VTEP consumes subnet string, and we get network from IP here, like `10.10.10.100/21` -> `10.10.8.0/21` (just example IP)

File: tests/network/bgp/evpn/libevpn.py [hunk 7]
```diff
@@ -468,3 +473,14 @@ def assert_evpn_workloads_connectivity(
...
+    host_cidrs = json.loads(node.instance.metadata.annotations["k8s.ovn.org/host-cidrs"])
+    return next(cidr for cidr in host_cidrs if "." in cidr and ipaddress.ip_network(cidr, strict=False).prefixlen < 32)
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: 1 -line generation usage, standard lib, nothing criminal

File: tests/network/bgp/evpn/libevpn.py [hunk 8]
```diff
@@ -1,12 +1,17 @@
...
 import logging
 import shlex
 import uuid
 from collections.abc import Generator
 from dataclasses import dataclass
+from typing import TYPE_CHECKING
 
 from ocp_resources.pod import Pod
+
+if TYPE_CHECKING:
```
Comment (Anatw) [inline tests/network/bgp/evpn/libevpn.py]: We usually have the if TYPE_CHECKING section after the import section. Is it placed here on purpose?

File: tests/network/bgp/evpn/libevpn.py (hunk 7 above)
Comment (Anatw) [inline tests/network/bgp/evpn/libevpn.py]: What if host-cidrs only contains /32 addresses? Then this raises a bare _StopIteration_, which is hard to debug - can we raise an explicit error with context instead?

File: tests/network/bgp/evpn/conftest.py [hunk 9]
```diff
@@ -88,9 +105,7 @@ def vtep(
...
```
Comment (qinqon) [inline tests/network/bgp/evpn/conftest.py]: this is the worker_cidr, the fact that is the vtep_cidr is something that can change in the future when for example they implement the managed vtep (loopback + new cidr) or we do that at nodes (since is the proper way of using vxlan).

File: tests/network/bgp/evpn/libevpn.py [hunk 10]
```diff
@@ -468,3 +473,14 @@ def assert_evpn_workloads_connectivity(
...
+
+
+def node_primary_ipv4_cidr(node: Node) -> str:
+    """Return the primary IPv4 CIDR of a node, skipping /32 host routes.
+
+    Parses k8s.ovn.org/host-cidrs and returns the first non-/32 IPv4
+    CIDR entry. On clusters where /32 host-route entries appear before
+    the interface CIDR, this ensures the correct address is returned.
+    """
+    host_cidrs = json.loads(node.instance.metadata.annotations["k8s.ovn.org/host-cidrs"])
```
Comment (qinqon) [inline tests/network/bgp/evpn/libevpn.py]: I have start the ovnk kind cluster with evpn activated and I get:

```yaml
k8s.ovn.org/host-cidrs: '["172.18.0.5/16","172.19.0.2/16","fc00:f853:ccd:e793::5/64"]'
```

There are two IPv4 addresses while at node status

```
status:
  addresses:
  - address: 172.18.0.5
    type: InternalIP
  - address: fc00:f853:ccd:e793::5
    type: InternalIP
```

We should read this annotation instead it also has sections per family and also you will not need to filter by /32 that btw can be a proper node IP for point to point or cloud setups.
```yaml
k8s.ovn.org/node-primary-ifaddr: '{"ipv4":"172.18.0.5/16","ipv6":"fc00:f853:ccd:e793::5/64"}'
```

So maybe some code like the following it will return directly the ip address type we can use later on 
```python
    node_primary_addr = json.loads(node.instance.metadata.annotations["k8s.ovn.org/node-primary-ifaddr"])
    return ipaddress.ip_interface(node_primary_addr["ipv4"])  
```

Then we can get the ip or cidr like 
```python
iface = node_primary_ipv4_cidr(node)
iface.ip       # 172.18.0.2
iface.network  # 172.18.0.0/16```

File: tests/network/bgp/evpn/libevpn.py (hunk 3 above)
Comment (qinqon) [reply tests/network/bgp/evpn/libevpn.py]: > our nodes are dual-stack. But we filter IPs via if "." in cidr. So, how it can return IPv6?

Well that's fragile, you can have ipv4 over ipv6 with `::ffff:172.18.0.2/96` 

I have propose python code to improve the situation

File: tests/network/libs/bgp.py
Comment (qinqon) [reply tests/network/libs/bgp.py]: Let's add this change so we know if customers need to add it or not (if they need to add we will see some failures).

but implement it with some comments pointing to the ovnk PRs.

**Review by qinqon — CHANGES_REQUESTED** [review]
We can really improve the node ip function

File: tests/network/bgp/evpn/conftest.py (hunk 9 above)
Comment (servolkov) [reply tests/network/bgp/evpn/conftest.py]: what is your concern\suggestion here?

File: tests/network/bgp/evpn/libevpn.py (hunk 8 above)
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: fixed, however if CI is green - I don't care.

File: tests/network/bgp/evpn/libevpn.py (hunk 3 above)
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: `::ffff:172.18.0.2/96` look like a very edge case, not sure we ever use this type of IP. However, I really like your suggestion with `node-primary-ifaddr` annotation, fixed the code to use it.

File: tests/network/bgp/evpn/libevpn.py (hunk 10 above)
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: `node-primary-ifaddr` - looks great, adapted the code to use it.

File: tests/network/bgp/evpn/libevpn.py (hunk 7 above)
Comment (servolkov) [reply tests/network/bgp/evpn/libevpn.py]: the helper has changed, pls take a look.

File: tests/network/libs/bgp.py
Comment (servolkov) [reply tests/network/libs/bgp.py]: fixed with dedicated commit

File: tests/network/libs/bgp.py [hunk 11]
```diff
@@ -214,20 +214,19 @@ def generate_frr_conf(
...
     lines.extend([
         " address-family ipv4 unicast",
         f"  network {external_subnet_ipv4}",
     ])
     for ip in nodes_ipv4_list:
         lines.extend([
             f"  neighbor {ip} activate",
-            f"  neighbor {ip} next-hop-self",
-            f"  neighbor {ip} route-reflector-client",
+            f"  neighbor {ip} nexthop-unchanged",
```
[the remark this answers is not in the document]
Comment (openshift-virtualization-qe-bot-3) [reply tests/network/libs/bgp.py]: ⚠️ This thread was automatically unresolved because it was resolved without a substantive response. Please address the review comment and explain how it was resolved before resolving this thread again.

**Comment by servolkov** [comment]
[Change: ](https://github.com/RedHatQE/openshift-virtualization-tests/compare/4f6583a605817abf93c7795291b59363b8237c9e..637a9b99a9be813ee56ede2ecf7aa3d56c6ce7a7) 
1. remove `set as-path exclude` option from frr.conf
2. refactor `node_primary_ipv4_cidr` -> `node_primary_ipv4_interface`

File: tests/network/libs/bgp.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/libs/bgp.py]: this is a very strange assertion. Based on this logic we should go over every single line and assert the config. Weird. Our main assertion that connectivity works over EVPN/BGP -> config is right.

**Review by qinqon — APPROVED** [review]
LGTM

File: tests/network/bgp/evpn/conftest.py (hunk 9 above)
Comment (servolkov) [reply tests/network/bgp/evpn/conftest.py]: resolving, the PR approved

**Comment by servolkov** [comment]
/verified

```
$ uv run pytest -m bgp --cluster-sanity-skip-nodes-check --cluster-sanity-skip-storage-check --skip-deprecated-api-test --cluster-sanity-skip-check
...
9 tests PASSED
```

**Comment by openshift-virtualization-qe-bot-3** [comment]
@coderabbitai
<details>
  <summary>Test execution plan request details</summary>

  CRITICAL: You MUST post an inline review comment on the first changed line of the first file.
  The inline comment should contain the full Test Execution Plan (smoke decision, gating decision, and specific affected tests).
  Do NOT submit a blocking review event (REQUEST_CHANGES/APPROVE).
  Post a single inline PR comment on Files Changed (non-blocking COMMENT flow).

  As an expert software testing engineer, analyze all modified files in this PR and create a targeted test execution plan.
  You will post an inline review comment with the test execution plan on the first changed file.
  If you fail to run or post a comment, retry.

  **Analysis Requirements:**

  1. Examine code changes in each modified file
  2. Identify affected code paths, functions, and classes
  3. Analyze pytest-specific elements: fixtures (scope, dependencies), parametrization, markers, conftest changes
  4. Trace test dependencies through imports, shared utilities, fixture inheritance, fixture teardown, and `yield from` cleanup in conftest
  5. Detect new tests introduced in the PR
  6. **Utilities and libs impact (when `utilities/` or `libs/` changes):**
     You MUST use shell scripts (`rg`, `git diff`) to trace the full impact.
     Follow these sub-steps in order:

     6a. **Identify modified symbols:** For each changed file under `utilities/` or `libs/`,
         list every modified function or method.
         Example: `git diff HEAD~1 --unified=0 -- utilities/hco.py | grep '^[+-]def '`

     6b. **Find direct callers:** Search tests and conftest for each symbol from 6a.
         Example: `rg -l 'get_hco_version' tests/`

     6c. **Trace fixture teardown and cleanup:** Find fixtures that reach
         the modified symbol through yield from or context-manager wrappers.
         Example: `rg -l 'yield from.*enable_common_boot|def.*enable_common_boot' tests/`

     6d. **Trace same-file callers:** In each changed file, find other functions
         whose body calls a modified symbol (including code after `yield`
         in `@contextmanager` helpers).
         Example: `rg 'get_hco_version|enable_common_boot' utilities/hco.py`

     6e. **Expand transitively:** If function A calls modified B, then
         tests/fixtures that call A are affected — even when the test body
         never imports B directly.

     Do NOT limit impact to tests that import the modified symbol only.
  7. **Smoke test impact:** Intersect the affected set from step 6 with smoke-marked tests.
     Run: `rg -l '@pytest.mark.smoke' tests/`
     VERIFY the above command returned actual file paths before concluding False.
     Set True if either condition is met:
     - a smoke-marked file appears in the affected set from 6b-6e, OR
     - any conftest.py in the smoke test's parent-directory hierarchy (up to repo root)
       imports or calls a modified utilities/libs symbol — including autouse fixtures
       that depend on modified functions. ALL tests in that directory and below are affected.
     Example check: for each smoke_file, scan dirname(smoke_file)/conftest.py,
     dirname(dirname(smoke_file))/conftest.py, etc. for modified symbol imports
     and autouse fixtures that depend on modified symbols.
  8. **Gating test impact:** Intersect the affected set from step 6 with gating-marked tests.
     Run: `rg -l '@pytest.mark.gating' tests/`
     Set True if a gating-marked file also appears in the affected set from 6b-6e.
     Utilities/libs changes often affect gating tests without affecting smoke tests.
     Do NOT stop analysis after concluding `Run smoke tests: False`.

  **Output rules:**
  Do NOT include analysis step numbers (1-8) in your visible output.

  **Your deliverable:**
  Your inline informational comment will be based on the following requirements:

  **Test Execution Plan**

  - **Run smoke tests: True / False** — If True, state the dependency path (test → fixture → changed symbol). True ONLY with a verified path.
  - **Run gating tests: True / False** — If True, state the dependency path. True if any gating-marked test is in the affected set.
  - **Affected tests to run** _(required when utilities/, libs/, or shared conftest changes — list concrete paths even when smoke is False)_

  Use these formats:
  - `path/to/test_file.py` - When the entire test file needs verification
  - `path/to/test_file.py::TestClass::test_method` - When specific test(s) needed
  - `path/to/test_file.py::test_function` - When specific test(s) needed
  - `-m marker` - When a marker covers multiple affected tests (e.g. `-m gating` only if ALL gating tests in scope need run)
  - Tag each listed test or group with its marker when not obvious, e.g. `(gating)` or `(smoke)`

  **Real test commands (MANDATORY when changes affect session/runtime code):**

  When the affected code runs at session/collection time (conftest fixtures, pytest plugins,
  config hooks, session-scoped setup) or modifies runtime behavior that unit tests mock away,
  you MUST include concrete `pytest` commands the PR author must run on a real cluster
  to verify the change works end-to-end. Include:
  - A command for the **error/fix path** (the scenario the PR fixes)
  - A command for the **happy path** (regression: the normal case still works)
  - Use lightweight tests (e.g., `--collect-only` for startup failures,
    a single small test for runtime behavior)
  If the PR only changes test logic (not utilities/libs/conftest), the affected test
  paths themselves serve as the real test commands — no separate section needed.

  Example output for a session-startup fix:
  ```
  **Real tests (cluster required)**
  Error path (the fix):
  `pytest tests/storage/.../test_foo.py --storage-class-matrix=nonexistent-sc --collect-only`
  Expected: ValueError with clear message, not IndexError

  Happy path (regression):
  `pytest tests/storage/.../test_foo.py --storage-class-matrix=<valid-sc> -k test_bar`
  Expected: session starts normally
  ```

  **Guidelines:**

  - Include tests affected directly OR via fixture setup/teardown, `yield from` cleanup, or transitive utility call chains (caller calls modified helper)
  - Use a full file path only if ALL tests in that file require verification
  - Use file path + test name when only specific tests use an affected fixture or utility wrapper (preferred for partial file impact)
  - If a test marker can cover multiple files/tests, provide the marker
  - Balance coverage vs over-testing - Keep descriptions minimal
  - Example: if leaf helper `foo()` changes, include tests whose fixture teardown calls wrapper `bar()` where `bar()` calls `foo()`, even when the test body only imports an unrelated symbol from the same utilities module

  **Hardware-Related Checks (SR-IOV, GPU, DPDK):**

  When PR modifies fixtures for hardware-specific resources:
  - **Collection Safety**: Fixtures MUST have existence checks (return `None` when hardware unavailable)
  - **Test Plan**: MUST verify both WITH and WITHOUT hardware:
    - Run affected tests on cluster WITH hardware
    - Verify collection succeeds on cluster WITHOUT hardware

  **CRITICAL WORKFLOW COMPLETION RULES:**

  When responding to this test execution plan request, you MUST follow these rules EXACTLY:

  1. **YOUR ONLY DELIVERABLE**: Post one non-blocking inline comment containing the test execution plan on the first changed line
  2. **THEN STOP IMMEDIATELY** - Do NOT generate any additional response
  3. **FALLBACK ONLY**: If inline comment API calls fail after retrying, post as a regular PR comment
  4. **SILENCE = SUCCESS**: After successfully submitting the review, your task is complete. No confirmation needed.

  **ABSOLUTE PROHIBITIONS** (violating these creates empty/meaningless reviews):
  - ❌ Do NOT post acknowledgment messages like "Test execution plan posted", "Review posted successfully", "I've successfully posted"
  - ❌ Do NOT mention review IDs, URLs, or confirmation of posting in the PR thread
  - ❌ Do NOT add any follow-up comments after submitting the review
  - ❌ Do NOT reply to confirm task completion
  - ❌ Do NOT explain what you did - just do it and stop

  **Remember**: The pull request review is visible to users. Additional comments are redundant noise.

  CRITICAL — Verification rules for smoke and gating decisions:
    - For each True decision, you must have traced a concrete path: test file → fixture → utility function → modified symbol.
    - Do NOT assume impact based on scope, file proximity, or infrastructure-sounding names.
      Only set True when you trace a concrete path through steps 6a-6e, then 7 or 8.
    WRONG: "utilities/hco.py changed → run smoke" (no traced dependency path)
    WRONG: "Run smoke tests: False" with no affected-test list when gating tests use the changed utility
    RIGHT: "Run smoke tests: False, Run gating tests: True" plus `tests/.../test_foo.py::TestBar::test_baz (gating)`
    RIGHT: "Smoke test X uses fixture Y → wrapper Z → changed function W in utilities/hco.py"

</details>

File: tests/network/bgp/evpn/conftest.py [hunk 12]
```diff
@@ -6,6 +6,7 @@
 import ocp_resources.network_operator_openshift_io as openshift_no
 import pytest
 from kubernetes.dynamic import DynamicClient
+from ocp_resources.cluster_operator import ClusterOperator
```
[the remark this answers is not in the document]
Comment (servolkov) [reply tests/network/bgp/evpn/conftest.py:9]: done https://github.com/RedHatQE/openshift-virtualization-tests/pull/6208#issuecomment-5540231847

**Review by Anatw — APPROVED** [review]
Thank you

**Review by vsibirsk — APPROVED** [review]
